### ISSUE TITLE
TypeError: browser.cdp is not a function

### ISSUE THREAD

**Issue by psmod2** [webdriverio/wdio-devtools-service#4]
I'm running the following test on the Selenium Grid 3.8.1 with Chrome v63.0.3239.132

However I keep seeing the error:
```
1) NETWORK TESTER should take JS coverage:
browser.cdp is not a function
running chrome
TypeError: browser.cdp is not a function
```

The test itself looks like:

```
it('Net Test', () => {
    browser.cdp('Network', 'enable')
    browser.on('Network.responseReceived', (params) => {
      console.log('======== ' + params.response.url + ' ============')
      console.log(params)
      console.log('====================')
    })

    HomePage.open()
  })
```

Any ideas?

**Comment by christian-bromann** [comment]
@psmod2 did you add `devtools` to your service list? Do you have a reproducible example ?

**Comment by psmod2** [comment]
@christian-bromann Yes, used `services: ['devtools']` in wdio.conf

**Comment by christian-bromann** [comment]
@psmod2 this should do the trick .. can you provide a reproducible example?

**Comment by psmod2** [comment]
A little difficult to provide a reproducible example - 
When on run this locally i.e. on my laptop, local Chrome it works fine.

I see the error only when running on my Selenium Grid - this is not public so not sure how to reproduce to you.

I've checked the npm list and I can see it is def installed:

├─┬ wdio-devtools-service@0.1.0
│ ├─┬ chrome-remote-interface@0.25.5
│ │ ├── commander@2.11.0
│ │ └─┬ ws@3.3.3
│ │   ├── async-limiter@1.0.0
│ │   ├── safe-buffer@5.1.1 deduped
│ │   └── ultron@1.1.1

The exact error I see is:

```
Couldn't find Chrome process
TypeError: Cannot read property 'cmd' of undefined
    at DevToolsService._callee2$ (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/wdio-devtools-service/build/devtools.js:153:82)
    at tryCatch (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/regenerator-runtime/runtime.js:62:40)
    at Generator.invoke [as _invoke] (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/regenerator-runtime/runtime.js:296:22)
    at Generator.prototype.(anonymous function) [as next] (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/regenerator-runtime/runtime.js:114:21)
    at step (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/babel-runtime/helpers/asyncToGenerator.js:17:30)
    at /mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/babel-runtime/helpers/asyncToGenerator.js:28:13
    at <anonymous>
F

0 passing (10.50s)
1 failing

1) Network Test should listen on network events:
browser.cdp is not a function
running chrome
TypeError: browser.cdp is not a function
    at Context.it (/mnt/mesos/sandbox/workspace/ati_x_develop-x/test/homepage/networkTest.js:7:13)
    at new Promise (<anonymous>)
    at new F (/mnt/mesos/sandbox/workspace/ati_x_develop-x/node_modules/core-js/library/modules/_export.js:35:28)
```

And my Grid looks like this:

![capture](https://user-images.githubusercontent.com/15928263/35543734-117fa62a-05aa-11e8-8315-2a9c81e8363a.PNG)

**Comment by christian-bromann** [comment]
Ahh there we go `Couldn't find Chrome process` .. looks like the service can't find the Chrome process. It is probably due to the environment you are running your tests in. Feel free to look into it and provide a PR with a fix.

**Comment by sns45** [comment]
To connect to remote host, you need a host ip and port number. 
I created a function which will accept both of these values:
```
browser.addCommand('cdpConnect', (host, num) => {
                this.chromePort = num;
                (async () => {
                    this.client = await this._getCDPClient(host, this.chromePort)
                    this.host = await this.client.host
                    this.port = await this.client.port
                    console.log("this host: " + this.client.host + " " + "port: "+ this.client.port)
                    /**
                     * propagate CDP events to the browser event listener
                     */
                    await this.client.on('event', (event) => browser.emit(event.method || 'event', event.params))
                })()
            })
```
From what I know, the port number is generated randomly by chrome driver(I guess). But it can be found out by navigating to ```chrome://version/```

```
browser.url("chrome://version/");
		portnum = actions.findcdpPort(browser.getText('#command_line')); //grabs the port number from commandline arguments present on the page  
		browser.cdpConnect(conf.config.host, portnum); //passing host ip and port to wdio-devtools-service
```

But this still didn't work and gave this error:
```
(node:32360) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 4): Error: connect ECONNREFUSED 1x2.2x.3x.xx:12949
(node:32360) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node
```
I guess chrome-remote-interface doesn't allow connection from outside. 
https://stackoverflow.com/questions/6827310/chrome-remote-debugging-doesnt-work-with-ip

We can create a proxy server on the remote machine to the relay the service outside of localhost. But I am not sure how it can be done if the port number is always randomly generated. Is there any way we can specify the ```--remote-debuggin-port=```` to something specific?

**Comment by christian-bromann** [comment]
This should be fixed with the latest version, can anyone approve this?

**Comment by psmod2** [comment]
Hi tried - Selenium Grid v3.8 - ChromeNode v64 and got:

```
Error: An element could not be located on the page using the given search parameters ("#command_line").
    at elements("#command_line") - getText.js:18:17
    at getText("#command_line") - index.js:353:27
    at url("chrome://version") - index.js:263:37
    at init() - runner.js:235:55
```
![untitled](https://user-images.githubusercontent.com/15928263/37014757-1d4e93e6-2146-11e8-9a01-1286de6a73d4.png)

**Comment by christian-bromann** [comment]
@psmod2 can you past the content of `chrome://version` in a gist and post it here?

**Comment by oroce** [comment]
I'm seeing the same problem.

[`chrome://version`](https://gist.github.com/oroce/f4ce2c6e5cbeecb85c76779418f1501e)

what I did to get it: 
```
browser.url('chrome://version');
console.log(browser.getSource());
```

As it turned out, if you specify `--remote-debugging-port=<portnum>` in `chromeOptions`'s args then the chromedriver will hang, since it also uses the remote debugging protocol. So there is no other way than trying to discover the port (see issue in chromedriver's bugtracker: https://bugs.chromium.org/p/chromedriver/issues/detail?id=878)

**Comment by christian-bromann** [comment]
> what I did to get it:

This is exactly what this service is doing.

**Comment by oroce** [comment]
@christian-bromann yeah I was aware of, but @sns45 came up with the idea that we should [define the port](https://github.com/webdriverio/wdio-devtools-service/issues/4#issuecomment-364241562) and I wanted to clarify that it's not possible.

Anyhow the `browser.cdp` is still `undefined`, can this happen whether `wdio` used in `sync` or `async` mode?

**Comment by christian-bromann** [comment]
> Anyhow the browser.cdp is still undefined, can this happen whether wdio used in sync or async mode?

No this is unrelated to that

**Comment by sns45** [comment]
@oroce yes you're right, after I've put that comment I did more research and I realized I was wrong.

As far as I know,  I guess it's a security feature of chrome that you can only connect via ```--remote-debugging-port```  if the host is localhost. It won't work if the host is anything other than localhost.
You cannot override the port number by providing it in the chromeoptions.

Going to ```chrome://version/``` will grab the port number to connect to that particular chrome instance locally so that multiple instances of chrome can be supported.

So, make sure you are running everything on local and also note that it may not work if you run it with gulp webdriverio launcher(that I don't know why).

**Comment by christian-bromann** [comment]
Thanks @sns45 for your feedback. I will close this then. If there is still an issue, I am happy to reopen.

**Comment by AnnikaNissen** [comment]
Trying to make the devtools-service work when running on a Selenium grid, I came across this issue.

@sns45  Does this 

> So, make sure you are running everything on local

mean, that devtools.service does not work when running on a Selenium Grid? Googling a bit, I came around [this](https://codereview.chromium.org/1920773003/) in the Chromium bugtracker, stating that you can set `--remote-debugging-address=0.0.0.0` when starting Chrome. Did you have a look into that? Any idea, what needs to be passt as `--remote-debugging-address`? The IP of the Selenium Grid hub, the IP of the computer running nodeJS/webdriverIO?

**Comment by christian-bromann** [comment]
It should not matter whether or not you run Chromedriver directly or via Selenium Grid.

**Comment by AnnikaNissen** [comment]
Unfortunately it does, at least for me. When running locally with 'selenium-standalone' server my code works perfectly. When running against a Selenium grid in my corporate network I get the following error, as soon as I have `services: ['devtools']` included in my wdio.config.js:

```
[12:53:53][Step 2/2] (node:472) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: connect ECONNREFUSED 127.0.0.1:12548
[12:53:53][Step 2/2] (node:472) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
[12:53:54][Step 2/2] (node:7200) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: connect ECONNREFUSED 127.0.0.1:12837
[12:53:54][Step 2/2] (node:7200) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

To get the error, I don't even need to use `browser.cdp(...)`, just including the service is enough. 

Any ideas? It would just be awesome to get this working!

Edit: by default I have multiple tests running in parallel on the grid, but I also tested with `maxInstances: 1` without seeing a difference.

**Comment by christian-bromann** [comment]
Looks like some connection issue. Hard to tell what the reason could be here. It is unrelated to the devtools service though and seems to be a general problem.

**Comment by mumming** [comment]
I am getting 
`UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 127.0.0.1:9222 `

when doing nothing to my tests but including 
` services: ['devtools'],`
in wdio.conf.js.

Did you find any solution for this @AnnikaNissen ?

**Comment by AnnikaNissen** [comment]
@mumming No - I didn't. Are you running your tests with a local selenium server or with a remote grid? 

I came to the conclusion, that using the devtools is not (or at least not our of the box) possible when running on a remote grid. Because:

The setup when using a grid looks like this:
```
| test.js |   -->  | Grid-hub |   -->   | Grid-node 1 - Chrome |   -->   | Webserver |
                                  -->   | Grid-node N - Chrome |   --> 
```

The commands of the devtools-service are executed, where test.js is executed, and the devtools-service tries to reach chrome on `localhost:XXXX`, where XXXX is the port read from the chrome://version/ page. That can't work, because the chrome is running on the grid node (--> a server in the www). 

As written earlier, there is a `--remote-debugging-address=0.0.0.0` option fro chrome. Maybe it is possible to get the setup to work with that option, but I didn't investigate it further (e.g. due to limited knowledge in networks). If you manage to get this to work, please your solution :)!

**Comment by christian-bromann** [comment]
@AnnikaNissen you made a good point here. The service doesn't work if Chrome is being run remotely. I think though that the service can be enhanced so that it can connect to remote Chrome browser.

**Comment by christian-bromann** [comment]
I will create a new issue.

**Comment by AnnikaNissen** [comment]
@christian-bromann Great! Would be awesome if it would work for remote-chrome too :)!

**Comment by mumming** [comment]
@AnnikaNissen Okay. I am running a local selenium server. I will post here if I find a solution :)

**Comment by mumming** [comment]
I also just end up on chrome://version/

**Comment by sns45** [comment]
@mumming 
You should run tests on local and
Notice the `--remote-debugging-port` number when it opens `chrome://version/`. If its set to `--remote-debugging-port=0` then you'll have to downgrade the `selenium-standalone` (the problem is probably with chrome-driver). It's working fine for me in `selenium-standalone version 6.13`.

This is a new issue caused by some external update(selenium-standalone or chrome-driver).

**Comment by CalderMarshall-zz** [comment]
@sns45 experiencing the same issue here,  adding ```['devtools']``` just opens ```chrome://version/``` which lists  ```--remote-debugging-port=0```

Without devtools listed as a service, error is ```browser.cdp is not a function```.

I am Using ```"selenium-standalone": "6.13.0"```, and testing locally.

**Comment by parthmakadiya12** [comment]
@CalderMarshall  same thing happening to me ..Do you find any luck ??
